Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-251.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: dcc911f0930dd7103a285c557b8ca7f8f677ce1d
Containerized:False
Number of Error(s): 5
src-react/components/forms/Fields.tsx(20,5): error TS2300: Duplicate identifier disabled.
src-react/components/forms/Fields.tsx(21,5): error TS2300: Duplicate identifier disabled.
make[2]: *** [esp/src/CMakeFiles/eclwatch.dir/build.make:311: esp/src/build/dist/stub.eclwatch.js] Error 1
make[1]: *** [CMakeFiles/Makefile2:10418: esp/src/CMakeFiles/eclwatch.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
Get ECLWatch build result
ReportTimes.
Finish at: 2021-08-31_10-14-40
time logs started
Prep time   :     8 sec (00:00:08)
Build time  :   166 sec (00:02:46)
Package time:     0 sec (00:00:00)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   174 sec (00:02:54)
time logs end

[Build error:](https://gist.github.com/ce042bdbdc824dfbad6bfece9c203a18)
Build log [Build log](https://gist.github.com/HPCCSmoketest/ce042bdbdc824dfbad6bfece9c203a18#file-result-21-08-31-10-11-32-log)

--- a/esp/src/src-react/components/DataPatterns.tsx
+++ b/esp/src/src-react/components/DataPatterns.tsx
@@ -1,0 +1,134 @@
+import * as React from "react";
+import { CommandBar, ContextualMenuItemType, ICommandBarItemProps, mergeStyleSets, ScrollablePane, ScrollbarVisibility, Sticky, StickyPositionType } from "@fluentui/react";
+import nlsHPCC from "src/nlsHPCC";
+import { DPWorkunit } from "src/DataPatterns/DPWorkunit";
+import { Report } from "src/DataPatterns/Report";
+import { useFile } from "../hooks/file";
+import { useWorkunit } from "../hooks/workunit";
+import { AutosizeHpccJSComponent } from "../layouts/HpccJSAdapter";
+import { Optimize } from "./forms/Optimize";
+import { ShortVerticalDivider } from "./Common";
+import { TargetClusterTextField } from "./forms/Fields";
+import { WUStatus } from "src/react/wuStatus";
+
+const dpStyles = mergeStyleSets({
+    inlineDropdown: {
+        marginTop: "6px",
+        ".ms-Dropdown": {
+            minWidth: "100px"
+        }
+    }
+});
+
+interface DataPatternsProps {
+    cluster: string;
+    logicalFile: string;
+}
+
+export const DataPatterns: React.FunctionComponent<DataPatternsProps> = ({
+    cluster,
+    logicalFile
+}) => {
+
+    const [file] = useFile(cluster, logicalFile);
+    const dpWu = React.useMemo(() => {
+        if (file) {
+            return new DPWorkunit(cluster, logicalFile, file?.Modified);
+        }
+        return undefined;
+    }, [cluster, file, logicalFile]);
+
+    const [showOptimize, setShowOptimize] = React.useState(false);
+    const [targetCluster, setTargetCluster] = React.useState<string>();
+    const [wuid, setWuid] = React.useState<string>();
+    const [wu, , , isComplete] = useWorkunit(wuid, true);
+
+    const refreshData = React.useCallback((full: boolean = false) => {
+        if (full) {
+            dpWu?.clearCache();
+        }
+        dpWu?.resolveWU().then(wu => {
+            setWuid(wu?.Wuid);
+        });
+    }, [dpWu]);
+
+    React.useEffect(() => refreshData(), [refreshData]);
+
+    const dpReport = React.useMemo(() => {
+        if (wu && isComplete) {
+            return new Report().wu(dpWu);
+        }
+    }, [wu, isComplete, dpWu]);
+
+    //  Command Bar  ---
+    const buttons = React.useMemo((): ICommandBarItemProps[] => [
+        {
+            key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
+            onClick: () => {
+                refreshData(true);
+            }
+        },
+        { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "analyze", text: nlsHPCC.Analyze, iconProps: { iconName: "AnalyticsView" },
+            disabled: !!wu,
+            onClick: () => {
+                dpWu?.create(targetCluster).then(() => {
+                    refreshData();
+                });
+            }
+        },
+        {
+            key: "targetCluster", itemType: ContextualMenuItemType.Normal,
+            commandBarButtonAs: () => <TargetClusterTextField
+                key="targetClusterField"
+                disabled={!!wu}
+                placeholder={nlsHPCC.Target}
+                className={dpStyles.inlineDropdown}
+                optional={false}
+                defaultSelectedKey={targetCluster}
+                onChange={(ev, row) => {
+                    setTargetCluster(row.key as string);
+                }}
+            />
+        }, {
+            key: "delete", text: nlsHPCC.Delete, iconProps: { iconName: "Delete" },
+            disabled: !isComplete,
+            onClick: () => {
+                dpWu?.delete().then(() => {
+                    setWuid("");
+                    refreshData();
+                });
+            }
+        },
+        { key: "divider_2", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "optimize", text: nlsHPCC.Optimize,
+            disabled: !isComplete,
+            onClick: () => setShowOptimize(true)
+        },
+        { key: "divider_3", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "wuid", text: wuid,
+            href: `#/workunits/${wuid}`,
+        }
+    ].filter(row => row.key !== "wuid" || !!wuid), [dpWu, isComplete, refreshData, targetCluster, wu, wuid]);
+
+    const rightButtons = React.useMemo((): ICommandBarItemProps[] => [
+    ], []);
+
+    return <ScrollablePane scrollbarVisibility={ScrollbarVisibility.auto}>
+        <Sticky stickyPosition={StickyPositionType.Header}>
+            <CommandBar items={buttons} overflowButtonProps={{}} farItems={rightButtons} />
+        </Sticky>
+        {wu?.isComplete() ?
+            <AutosizeHpccJSComponent widget={dpReport} /> :
+            wuid ? <div style={{ width: "512px", height: "64px", float: "right" }}>
+                <WUStatus wuid={wuid}></WUStatus>
+            </div> :
+                <h3>{nlsHPCC.DataPatternsNotStarted}</h3>
+        }
+        <Optimize dpWu={dpWu} targetCluster={targetCluster} logicalFile={logicalFile} showForm={showOptimize} setShowForm={setShowOptimize} />
+    </ScrollablePane>;
+
+};

--- a/esp/src/src-react/components/FileDetails.tsx
+++ b/esp/src/src-react/components/FileDetails.tsx
@@ -18,6 +18,7 @@ import { ShortVerticalDivider } from "./Common";
 import { FileDetailsGraph } from "./FileDetailsGraph";
 import { TableGroup } from "./forms/Groups";
 import { CopyFile } from "./forms/CopyFile";
+import { DataPatterns } from "./DataPatterns";
 import { DesprayFile } from "./forms/DesprayFile";
 import { RenameFile } from "./forms/RenameFile";
 import { ReplicateFile } from "./forms/ReplicateFile";
@@ -207,6 +208,7 @@ export const FileDetails: React.FunctionComponent<FileDetailsProps> = ({
                     <Result cluster={cluster} logicalFile={logicalFile} />
                 </PivotItem>
                 <PivotItem headerText={nlsHPCC.DataPatterns} itemKey="DataPatterns" style={pivotItemStyle(size, 0)}>
+                    <DataPatterns cluster={cluster} logicalFile={logicalFile} />
                 </PivotItem>
                 <PivotItem headerText={nlsHPCC.ECL} itemKey="ECL" style={pivotItemStyle(size, 0)}>
                     <ECLSourceEditor text={file?.Ecl} readonly={true} />

--- a/esp/src/src-react/components/forms/Fields.tsx
+++ b/esp/src/src-react/components/forms/Fields.tsx
@@ -61,7 +61,7 @@ const Dropdown: React.FunctionComponent<DropdownProps> = ({
         setSelOptions(selOptions);
     }, [optional, options, defaultSelectedKey, handleOnChange]);
 
-    return <DropdownBase key={key} label={label} errorMessage={errorMessage} required={required} className={className} defaultSelectedKey={defaultSelectedKey} selectedKey={selectedKey} onChange={handleOnChange} placeholder={placeholder} disabled={disabled} options={selOptions} />;
+    return <DropdownBase key={key} label={label} errorMessage={errorMessage} required={required} className={className} defaultSelectedKey={defaultSelectedKey} selectedKey={selectedKey} onChange={handleOnChange} placeholder={placeholder} options={selOptions} disabled={disabled} />;
 };
 
 export type FieldType = "string" | "number" | "checkbox" | "datetime" | "dropdown" | "link" | "links" | "progress" |

--- a/esp/src/src-react/components/forms/Optimize.tsx
+++ b/esp/src/src-react/components/forms/Optimize.tsx
@@ -1,0 +1,108 @@
+import * as React from "react";
+import { Checkbox, DefaultButton, PrimaryButton, TextField, } from "@fluentui/react";
+import { scopedLogger } from "@hpcc-js/util";
+import { useForm, Controller } from "react-hook-form";
+import nlsHPCC from "src/nlsHPCC";
+import { MessageBox } from "../../layouts/MessageBox";
+import { TargetClusterTextField } from "./Fields";
+import { DPWorkunit } from "src/DataPatterns/DPWorkunit";
+import { pushUrl } from "src-react/util/history";
+
+const logger = scopedLogger("../components/forms/PublishQuery.tsx");
+
+interface PublishFormValues {
+    target: string;
+    logicalFile: string;
+    overwrite: boolean;
+}
+
+interface OptimizeProps {
+    dpWu: DPWorkunit;
+    targetCluster: string;
+    logicalFile: string;
+
+    showForm: boolean;
+    setShowForm: (_: boolean) => void;
+}
+
+export const Optimize: React.FunctionComponent<OptimizeProps> = ({
+    dpWu,
+    targetCluster,
+    logicalFile,
+    showForm,
+    setShowForm
+}) => {
+
+    const defaultValues: PublishFormValues = React.useMemo(() => ({
+        target: targetCluster,
+        logicalFile: `${logicalFile}::optimized`,
+        overwrite: false
+    }), [logicalFile, targetCluster]);
+
+    const { handleSubmit, control } = useForm<PublishFormValues>({ defaultValues });
+
+    const closeForm = React.useCallback(() => {
+        setShowForm(false);
+    }, [setShowForm]);
+
+    const onSubmit = React.useCallback(() => {
+        handleSubmit(
+            (data, evt) => {
+                dpWu.optimize(data.target, data.logicalFile, data.overwrite).then(wu => {
+                    pushUrl(`/workunits/${wu.Wuid}`);
+                }).catch(logger.error);
+            },
+            logger.info
+        )();
+    }, [dpWu, handleSubmit]);
+
+    return <MessageBox show={showForm} setShow={closeForm} title={nlsHPCC.Optimize}
+        footer={<>
+            <PrimaryButton text={nlsHPCC.Optimize} onClick={handleSubmit(onSubmit)} />
+            <DefaultButton text={nlsHPCC.Cancel} onClick={() => closeForm()} />
+        </>}>
+        <Controller control={control} name="target" render={({
+            field: { onChange, name: fieldName, value },
+            fieldState: { error }
+        }) => <TargetClusterTextField
+                key="targetClusterField"
+                label={nlsHPCC.Target}
+                placeholder={nlsHPCC.Target}
+                defaultSelectedKey={value}
+                optional={false}
+                onChange={(ev, row) => {
+                    return onChange(row.key);
+                }}
+                errorMessage={error && error?.message}
+            />}
+            rules={{
+                required: nlsHPCC.ValidationErrorRequired
+            }}
+        />
+        <Controller
+            control={control} name="logicalFile"
+            render={({
+                field: { onChange, name: fieldName, value },
+                fieldState: { error }
+            }) => <TextField
+                    name={fieldName}
+                    onChange={onChange}
+                    label={nlsHPCC.Name}
+                    value={value}
+                    errorMessage={error && error?.message}
+                    styles={{ root: { minWidth: 320 } }}
+                />}
+            rules={{
+                required: nlsHPCC.ValidationErrorRequired
+            }}
+        />
+        <div style={{ paddingTop: "10px" }}>
+            <Controller
+                control={control} name="overwrite"
+                render={({
+                    field: { onChange, name: fieldName, value }
+                }) => <Checkbox label={nlsHPCC.Overwrite} name={fieldName} checked={value} onChange={onChange} />}
+            />
+        </div>
+    </MessageBox>;
+};

--- a/esp/src/src/ESPRequest.ts
+++ b/esp/src/src/ESPRequest.ts
@@ -133,7 +133,7 @@ class RequestHelper {
 
     send(service, action, params?): Promise<any> {
         if (!this.isSessionCall(service, action) && (!this.hasServerSetCookie() || (this.hasAuthentication() && !this.isAuthenticated()))) {
-            // tslint:disable-next-line: deprecation
+            // @ts-ignore
             window.location.reload(true);
             return new Promise((resolve, reject) => { });
         }

--- a/esp/src/src/Utility.ts
+++ b/esp/src/src/Utility.ts
@@ -292,7 +292,9 @@ export function downloadToCSV(grid, rows, fileName) {
         const a = document.createElement("a");
         mimeType = mimeType || "application/octet-stream";
 
+        // @ts-ignore
         if (navigator.msSaveBlob) { // IE10
+            // @ts-ignore
             return navigator.msSaveBlob(new Blob([content], { type: mimeType }), fileName);
         } else if ("download" in a) {
             a.href = "data:" + mimeType + "," + encodeURIComponent(content);
@@ -436,7 +438,7 @@ export function alphanumCase(a, b) {
 
 export function onDomMutate(domNode, callback, observerOpts) {
     observerOpts = observerOpts || { attributes: true, attributeFilter: ["style"] };
-    var observer = new MutationObserver(mutations => {
+    const observer = new MutationObserver(mutations => {
         if (domNode.offsetParent === null) return;
         observer.disconnect();
         if (typeof callback === "function") {

